x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/RoleDescriptorStore.java

```diff
@@ -239,11 +241,27 @@ private void loadRoleDescriptorsAsync(Set<String> roleNames, ActionListener<Role
                         roleNames.remove(descriptor.getName());
                     }
                 } else {
-                    logger.warn(new ParameterizedMessage("role retrieval failed from [{}]", rolesProvider), result.getFailure());
+                    logger.warn(
+                        new ParameterizedMessage("role [{}] retrieval failed from [{}]", roleNames, rolesProvider),
+                        result.getFailure()
+                    );
                     rolesResult.setFailure();
                 }
                 providerListener.onResponse(result);
-            }, providerListener::onFailure));
+            }, e -> {
+                // If we're resolving a set that includes the `superuser` role, then ignore failures for subsequent roles
+                if (isSuperuserRetrieval && containsSuperuser(roleNames) == false) {
+                    logger.warn(new ParameterizedMessage("role [{}] retrieval failed from [{}]", roleNames, rolesProvider), e);
+                    rolesResult.setFailure();
+                    providerListener.onResponse(null);
```

Comment:
A few comments:
1. Is it necessary to check `containsSuperuser` again? The `ReservedRoleStore` is first in place to resolve the role names and is guaranteed to resolve the `superuser` role.
2. I assume the special handling for superuser here is to help BWC because existing code short circuits on `superuser` role and does not even attempt to resolve any other role names. Without this logic, some existing users may suddenly stop working because they may have a role name that causes error on resolving. But on the other hand, failure of `RolesRetrievalResult` prevents the role from being cached which could lead to subtle and hard to pin down performance degradation. Also what does this mean in the future, i.e.`superuser` role is no longer short circuit but it still has a special effect on role retrieving for any roles defined in the future? In short, I wonder whether this is a form of leniency that we should avoid.

Reply:
It's not strictly necessary, but if we remove it then we are implicitly assuming that `ReservedRoleStore` never fails. I don't like building that assumption in here.

Reply:
The behaviour is driven from the test.

We want to be able to guarantee that users with `superuser` role can admin the cluster even something goes wrong with retrieve other incidental roles. And due to the way role mapping works, it's quite possible to have other indicendal roles.

We previously guaranteed that by relying on that fact that `superuser` was all powerful, so there was no need to resolve any other roles. That's no longer true, so we needed to remove that special short circuit.  
However, the consequence is that a user with `superuser` might be prevented from performing any operations because of a problem retrieving their incidental roles, which isn't helpful.

It's a form of leniency I guess, but only in the sense of "make it possible to administer a cluster even if some things are broken".

Reply:
> make it possible to administer a cluster even if some things are broken

This is a good point and thanks for the explanation. Along this line of thought, I wonder this should be handled in `CompositeRolesStore#buildRoleFromRoleReference` which can even cater for `ReservedRoleStore` failure and be closer to today's behaviour. So something like:

```
roleReference.resolve(roleReferenceResolver, ActionListener.wrap(rolesRetrievalResult -> { ... }, 
  e -> {
     if (roleReference has superuser role) {
       listener.onResponse(superuserRole);
     }
  }
);
```

If above change is too much of a hassle for testing, I am OK for you to skip it since `ReservedRoleStore` failure is rather unlikely.

Reply:
Yeah, that sounds like it could be a neater solution. Let me see how it goes.